cpp/src/cylon/indexing/slice.cpp

```diff
@@ -0,0 +1,169 @@
+     * @param table, offset and length
+     * @return new sliced table
+     */
+
+
+    Status LocalSlice(const std::shared_ptr<Table> &in, int64_t offset, int64_t length,
+                std::shared_ptr<cylon::Table> &out) {
+    const auto &ctx = in->GetContext();
+    std::shared_ptr<arrow::Table> out_table, in_table = in->get_table();
+
+    if (!in->Empty()) {
+        out_table = in_table->Slice(offset, length);
+    } else {
+        out_table = in_table;
+    }
+    return Table::FromArrowTable(ctx, std::move(out_table), out);
+    }
+
+
+    /**
+     * DistributedSlice the part of table to create a single table
+     * @param table, offset and length
+     * @return new sliced table
+     */
+
+
+    Status DistributedSlice(const std::shared_ptr<cylon::Table> &in, int64_t offset, int64_t length,
+                std::shared_ptr<cylon::Table> &out) {
+
+    const auto &ctx = in->GetContext();
+    std::shared_ptr<arrow::Table> out_table, in_table = in->get_table();
+    auto num_row = in->Rows();
+    
+    std::vector<int64_t> sizes;
+    std::shared_ptr<cylon::Column> sizes_cols;
+    RETURN_CYLON_STATUS_IF_FAILED(Column::FromVector(sizes, sizes_cols));
+
+    auto num_row_scalar = std::make_shared<Scalar>(arrow::MakeScalar(num_row));
+    
+
+    RETURN_CYLON_STATUS_IF_FAILED(ctx->GetCommunicator()->Allgather(num_row_scalar, &sizes_cols));
+
+    auto *data_ptr =
+        std::static_pointer_cast<arrow::Int64Array>(sizes_cols->data())
+            ->raw_values();
+
+    int64_t L = length;
+    int64_t K = offset;
+    int64_t zero_0 = 0;
+    int64_t rank = ctx->GetRank();
+    int64_t L_i = std::accumulate(data_ptr, data_ptr + rank, zero_0);
+
+    int64_t sl_i = *(data_ptr + rank);
+
+
+    int64_t x = std::max(zero_0, std::min(K - L_i, sl_i));
+    int64_t y = std::min(sl_i, std::max(K + L - L_i, zero_0)) - x;
+
+
+    out_table = in_table->Slice(x, y);
+
+    return Table::FromArrowTable(ctx, std::move(out_table), out);
+    }
+
+
+    /**
+     * Head the part of table to create a single table with specific number of rows
+     * @param tables, number of rows
+     * @return new table
+     */
+
+    Status Head(const std::shared_ptr<Table> &table, int64_t num_rows, std::shared_ptr<cylon::Table> &output) {
+
+    std::shared_ptr<arrow::Table>  in_table = table->get_table();
+    const int64_t table_size = in_table->num_rows();
+
+    if(num_rows > 0 && table_size > 0) {
+        return LocalSlice(table, 0, num_rows, output);
+    }
+    else
+        return cylon::Status(Code::IOError);
+
+    }
+
+    Status DistributedHead(const std::shared_ptr<Table> &table, int64_t num_rows, std::shared_ptr<cylon::Table> &output) {
+
+    std::shared_ptr<arrow::Table>  in_table = table->get_table();
+    const int64_t table_size = in_table->num_rows();
+
+    if(num_rows > 0 && table_size > 0) {
+        return DistributedSlice(table, 0, num_rows, output);
+    }
+    else
+        return cylon::Status(Code::IOError);
```

Comment:
need to add an appropriate error message here!

Reply:
done